### ISSUE TITLE
Firebase Adapter Fails with App Name Already exist error

### ISSUE THREAD

**Issue by satish860** [nextauthjs/next-auth#4927]
### Adapter type

@next-auth/firebase-adapter

### Environment

yarn dev

### Reproduction URL

https://github.com/satish860/codegist

### Describe the issue

I am getting the following error. 

`FirebaseError: Firebase: Firebase App named '[DEFAULT]' already exists with different options or config (app/duplicate-app).`

![image](https://user-images.githubusercontent.com/817360/178895566-c45bc524-6108-4a6a-acbe-a6168f86635f.png)

In my initial debugging, I found Firebase adapter is trying to initialize the app, which is already initialized. 

What if we can add the similar code 
`// Initialize Firebase
const app = !getApps().length ? initializeApp(firebaseConfig) : getApp();`

### How to reproduce

Step 1:  Run the app in the repo using Yarn dev
Step 2: Click the Sign link
Step 3: You will be getting the following auth error screen

![image](https://user-images.githubusercontent.com/817360/178896099-387195b1-e132-4da5-9b9f-49a170d5938e.png)

### Expected behavior

It should redirect me to the Provider Login screen

**Comment by balazsorban44** [comment]
@chanceaclark let's revisit and allow passing an app, as people might have already set up one outside next-auth. 👍

**Comment by ifeoluwaayo** [comment]
Going through this same issue please kindly fix or if there's a workaround I would love to know.

**Comment by Chibionos** [comment]
Any scope of passing an firestore object created from firebase-admin ? that would solve our security rules issues #5049

**Comment by Chibionos** [comment]
@ifeoluwaayo how are you setting up your security rules, I am curious as I see I cannot open the security rules for the collections this adapters requires. Did you find a work around ?

**Comment by chanceaclark** [comment]
So I think there is a big gotcha here: **If you initialize a firebase app on the client, you need to initialize a new one on the server as well.**

I went ahead and added some logic in #5055 to make sure we get the default app or initialize a new one to prevent that duplicate app issue. Along with that I also added some documentation about reusing your client app inside the server config.

**Comment by 27lakshay** [comment]
I'm facing this same issue, I hope this gets merged quickly

**Comment by itsmylife** [comment]
@27lakshay @chanceaclark I am waiting for that [PR](https://github.com/nextauthjs/next-auth/pull/5055) too. It's been very long and I can imagine the workload that people have. So let's cross our fingers 🤞 that will have some attention soon.

**Comment by Sanjay-B** [comment]
The reason this is happening is because, you're trying to override an instance that has no identifier given. The docs for this is extremely confusion but, here's something to think about:

```js
import { initializeApp } from 'firebase/app';
const firebaseApp = initializeApp({ ... });
```
If you did something like this previously somewhere in your codebase and it was executed before the adapter was able to create an instance, it'll default to `"[DEFAULT]"`. There's a second parameter that's optional that allows you to have *multiple* instances with an identifier you provide. i.e `initializeApp({ ... }, identifierString)`. A simple solution to this could be something like, `initializeApp({ ... }, "NEXT_AUTH")`. 

There's a couple things wrong with the adapter that should probably be corrected sometime soon. You shouldn't be initializing a new default app. You should be able to pass an instanced app so that there's some obvious source of truth that is visible and fits different use-cases. From a developer standpoint, when firebase errors like it does and you have multiple instances of firebase active on the app, you literally have to go through each one to figure out which one errored. To avoid that, it should be clear. An adapter is meant to provide an interface between NEXT_AUTH datatypes and firebase so there's a compatibility layer between both. It should not be instancing things. I can fix this myself because, I ran into this error, took me a couple hours to figure out it was this and now I am here.

Secondly, [firebase adapter] - Firebase adapter requires open security rules we cannot allow in production (https://github.com/nextauthjs/next-auth/issues/5049) exposes a very insecure approach that is actively being pitched as a possible solution. This is totally incorrect. Having insecure database rules where access is not controlled makes a lot of this a no-opt solution when it comes to production ready environments.

There's probably a solid amount of users that use this adapter with open security rules without actually realizing it and that puts user-data at risk if you're not in a development environment. It's a small problem if you have small users. It starts to multiply as your user-base grows.

**Comment by drunkenvalley** [comment]
It seems to me that Firebase wants you to get the existing app.

In my own adapter for Firebase admin SDK I experienced this issue, and I wound up implementing a fix:

```typescript
import { initializeApp, getApp, getApps } from 'firebase-admin/app'

const appList = getApps()
const appName = 'firebase-admin-adapter' // arbitrary name
const app = !!appList.length && !!appList.find(a => a.name === appName)
    ? getApp(appName)
    : initializeApp(options, appName)
```

I see the current firebase adapter initializes the app without verifying whether one already exists, causing problems.

**Comment by bcthakre** [comment]
@drunkenvalley . Are you using the below code on firebase.config.js or [...nextauth].js

`import { initializeApp, getApp, getApps } from 'firebase-admin/app'

const appList = getApps()
const appName = 'firebase-admin-adapter' // arbitrary name
const app = !!appList.length && !!appList.find(a => a.name === appName)
    ? getApp(appName)
    : initializeApp(options, appName)`

**Comment by bcthakre** [comment]
Thanks Balazs. Will there be a documentation on how to update current setup. Also, what changes need to be done on firestore security rules. Best Regards, Bhupendrasinh ThakreOn Feb 5, 2023, at 7:33 AM, Balázs Orbán ***@***.***> wrote:﻿
Closed #4927 as completed via 6445eb6.

**Comment by PigeonZow** [comment]
Will the update be posted in this thread? I'm running into the same issue, not using firebase anywhere except for in [...nextauth].ts
